### PR TITLE
Reorganize file utils

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR reorganizes the gigantic `file_utils` in several submodules, namely:
- `doc_utils` contain all the decorators we use for our docstrings as well as the generic examples
- `file_utils` keep all Hub-related functionality (it might warrant a rename to `hub_utils` see below)
- `generic_utils` contain the "real" utils like custom Enums, to_py_obj etc.
- `import_utils` contains all the `is_xxx_available` as well as the modules that power our lazy inits

To avoid any breaking change, this is all done while reimporting the objects moved into `file_utils`. The only thing that will break is if a user imported a private object from there that has moved, but I think that's okay. The fact all tests pass while no import (except the private objects) was changed (yet) proves it.

The easiest way to review this PR is to just look at the `file_utils` module and look after my comment, where you will find a summary of the functions/classes/constants that were moved in the form of imports. The rest of the diff (apart from the main util files) come from private objects imports, and one import that was incorrect.

I have kept the PR simple for now but will change all the relative imports in every file to take each object from its real source, but before that I wanted to have your opinion on the splits done, as well as the two following points:
- as said above, we could also rename `file_utils` to a more appropriate `hub_utils` (while keeping a `file_utils` for backward compatibility but with just the imports).
- also, the modules I have created could also go in the utils subfolder instead of the root subfolder (and be named `.utils.doc`, `.utils.hub`, `.utils.generic` and `.utils.import` for instance), still with keeping BC.

As soon as we all agree on the move and the names of the new submodules, I will make the changes of import paths in all the files of the lib :-)

### CODE REVIEW

File: src/transformers/utils/fx.py [hunk 1]
```diff
@@ -45,8 +45,9 @@
     XLNetForQuestionAnswering,
     logging,
 )
-from ..file_utils import TORCH_FX_REQUIRED_VERSION, importlib_metadata, is_torch_fx_available
+from ..file_utils import TORCH_FX_REQUIRED_VERSION, is_torch_fx_available
 from ..models.auto import get_values
+from ..utils.versions import importlib_metadata
```
Comment (sgugger) [inline src/transformers/utils/fx.py:50]: This was imported from the wrong place.

File: src/transformers/file_utils.py [hunk 2]
```diff
@@ -12,1768 +12,192 @@
...
+from requests.exceptions import HTTPError
+from transformers.utils.logging import tqdm
 
-    ```python
-    >>> from transformers import {processor_class}, {model_class}
+from . import __version__
 
-    >>> tokenizer = {processor_class}.from_pretrained("{checkpoint}")
-    >>> model = {model_class}.from_pretrained("{checkpoint}")
+# Backward compatibility imports, to make sure all those objects can be found in file_utils
```
Comment (sgugger) [inline src/transformers/file_utils.py:45]: From here to line 122 is what you should read to review this PR, no code has change and this summarizes all the moves done.

File: src/transformers/__init__.py [hunk 3]
```diff
@@ -91,6 +91,7 @@
     "debug_utils": [],
     "dependency_versions_check": [],
     "dependency_versions_table": [],
+    "doc_utils": [],
```
Comment (sgugger) [inline src/transformers/__init__.py:94]: Additions here are for the repo consistency checks to pass. When we settle on names and objects to be moved, each submodule will contain the objects of the public init it defines.

**Comment by gante** [comment]
> also, the modules I have created could also go in the utils subfolder instead of the root subfolder (and be named .utils.doc, .utils.hub, .utils.generic and .utils.import for instance), still with keeping BC.

Related question: we have many files named `xxx_utils.py` outside the `utils` folder. Is it due to legacy files, or is it intentional?

File: src/transformers/file_utils.py [hunk 4]
```diff
@@ -12,1768 +12,192 @@
...
-    >>> inputs = tokenizer("Hello, my dog is cute", return_tensors="np")
-    >>> outputs = model(**inputs)
+SENTENCEPIECE_UNDERLINE = "▁"
+SPIECE_UNDERLINE = SENTENCEPIECE_UNDERLINE  # Kept for backward compatibility
 
-    >>> # retrieve logts for next token
-    >>> next_token_logits = outputs.logits[:, -1]
-    ```
-"""
+MULTIPLE_CHOICE_DUMMY_INPUTS = [
```
Comment (patrickvonplaten) [inline src/transformers/file_utils.py:179]: Are those constants imported a lot? They are only used for TF models no? Should we maybe move them to `modeling_tf_utils.py`?

**Comment by patrickvonplaten** [comment]
Thanks a lot for working on this! Very much needed change :-) 

Regarding the two points above:

- I'd be in favor of keep calling the file `file_utils.py` actually as I think it's not **only** related to the Hub. E.g. the name of the model weights is also how they will be called locally "pytorch_model.bin". So don't think it's worth renaming the file here
- I would be in favor of moving everything into a utils folder actually

File: src/transformers/doc_utils.py [hunk 5]
```diff
@@ -0,0 +1,936 @@
...
+    >>> tokenizer = {processor_class}.from_pretrained("{checkpoint}")
+    >>> model = {model_class}.from_pretrained("{checkpoint}")
+
+    >>> inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
+    >>> labels = torch.tensor([1] * inputs["input_ids"].size(1)).unsqueeze(0)  # Batch size 1
+
+    >>> outputs = model(**inputs, labels=labels)
+    >>> loss = outputs.loss
+    >>> logits = outputs.logits
+    ```
```
Comment (LysandreJik) [inline src/transformers/doc_utils.py:165]: Maybe at some point we could provide a better example than this :smile:

**Review by LysandreJik — APPROVED** [review]
That file really needed to be split, indeed. Agree with both points as well (keeping it named `file_utils.py` and moving everything in the `utils/` directory).

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for working on this! I'm in favour of creating a utils subfolder.

**Comment by sgugger** [comment]
As per your comments, I've put everything in a `utils` submodule and reimported everything:
- in `file_utils` for backward compatibility (again see all tests passing with no change yet)
- in the `utils` init: this way if we deprecate `file_utils` at some point, the code migration is super easy: replace `from transfomers.file_utils import xxx` by `from transformers.utils import xxx`.

If you agree, I'll clean up all internal imports then merge @patil-suraj @patrickvonplaten @LysandreJik

**Comment by sgugger** [comment]
> Related question: we have many files named xxx_utils.py outside the utils folder. Is it due to legacy files, or is it intentional?

@gante This is more or less legacy from where the repo was flat and had no subfolders. Some could probably go in the utils subfolder as well, though I'm keeping this PR focused on `file_utils` (it will be big enou already ;-) )

**Review by patrickvonplaten — APPROVED** [review]
